8217472: Add attenuation for PointLight

nlisker opened the pull request:
CSR: https://bugs.openjdk.java.net/browse/JDK-8218264

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8217472](https://bugs.openjdk.java.net/browse/JDK-8217472): Add attenuation for PointLight


### Reviewers
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/43/head:pull/43`
`$ git checkout pull/43`

nlisker commented:
Tested D3D on Win 10 and GL on Ubuntu 18.04.

There is still a bug (at least on Win) for some certain combinations of values as shown in this picture:
![image](https://user-images.githubusercontent.com/37422899/69054658-c2ad9180-0a14-11ea-809c-15c8ae485fd0.png)

This sample program can be used to test the changes.
Controls: mouse wheel zooms, RMB drag rotates, LMB pans.

[sample.zip](https://github.com/openjdk/jfx/files/3858700/sample.zip)

nlisker commented:
Kevin, Ambarish,

You can start the review, especially the API. I will hunt that specific values bug this week.

I'll need to know what kind of tests are needed in terms of functionality and performance.

nlisker commented:
The bug I mentioned above is not a bug actually. There are rapid changes over a small distance that make it looks like a jump in the lighting values, but when working with a finer scale the lighting dynamics seem correct.

kevinrushforth commented:
I think this is on the right track. The API looks like it is in good shape.

This will need a fair bit of testing to ensure that there are no regressions either in functionality or (especially) performance, in addition to tests for the new functionality. On the performance aspect, the inner loop of the lighting calculation now has an additional if test for the max range and additional arithmetic calculations for the attenuation. What we will need is a test program that we can run on Mac and Windows to measure the performance of rendering in a fill-rate-limited case. Ideally, we would not pay much of a performance hit in the default case where `ca == 1, la == 0, qa == 0`, but we first need to be able to measure the drop in performance before we can say whether it is acceptable.

Speaking of testing, I took the current patch for a test drive on Mac and Windows. I get the following system test failures on Mac, and also the same failure using fx83dfeatures/LightMotion in toys.


```
Shader compile log: ERROR: 0:308: Use of undeclared identifier 'range'
ERROR: 0:316: Regular non-array variable 'dist' may not be redeclared

test.robot.test3d.MeshCompareTest > testSnapshot3D[3] STANDARD_ERROR
    java.lang.RuntimeException: Error creating fragment shader
    	at javafx.graphics/com.sun.prism.es2.ES2Shader.createFromSource(ES2Shader.java:141)
    	at javafx.graphics/com.sun.prism.es2.ES2PhongShader.getShader(ES2PhongShader.java:177)
        ...
test.robot.test3d.MeshCompareTest > testSnapshot3D[3] FAILED
    java.lang.IllegalArgumentException: Unrecognized image loader: null
        at javafx.graphics/javafx.scene.image.WritableImage.loadTkImage(WritableImage.java:278)
        at javafx.graphics/javafx.scene.image.WritableImage$1.loadTkImage(WritableImage.java:53)
        at javafx.graphics/javafx.scene.Scene.doSnapshot(Scene.java:1340)
        at javafx.graphics/javafx.scene.Scene.doSnapshot(Scene.java:1372)
        at javafx.graphics/javafx.scene.Scene.snapshot(Scene.java:1462)
        at test.robot.test3d.MeshCompareTest.lambda$testSnapshot3D$0(MeshCompareTest.java:315)


test.robot.test3d.Snapshot3DTest > testSnapshot3D[3] FAILED
(same failure as above)


test.robot.test3d.Snapshot3DTest > testSnapshot3D[7] FAILED
(same failure as above)
```

nlisker commented:
> I get the following system test failures on Mac
> ```
> Shader compile log: ERROR: 0:308: Use of undeclared identifier 'range'
> ERROR: 0:316: Regular non-array variable 'dist' may not be redeclared
> ```

I don't have a Mac to test on, but on Ubuntu system tests pass (I ran the `test` command for systemTests). Does the sample app I attached also fail on Mac? They both use the same shaders, so I wonder where the issue could be.
Moreover, the error messages are strange. `dist` is not redeclared and `range` is not undeclared in the shader. The error message seems to originate from the native function `Java_com_sun_prism_es2_GLContext_nCompileShader` in `GLContext.c`, not managing to compile the shader, but I can't tell why.

kevinrushforth commented:
I get the same error on Ubuntu 16.04 as on Mac. Did you run the system tests with `-PFULL_TEST=true -PUSE_ROBOT=true`? Also, you can try running the `fx83dfeatures/LightMotion` toy and you should see the same error.

I still need to test your sample app on Mac.

arapte commented on modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGPointLight.java:
```diff
@@ -26,11 +26,85 @@
+    }
+
+
+    private double ca = DEFAULT_CA;
```
The coefficients are not directly used in any arithmetic on java side. They are converted to `float` and passed to GL or D3D pipeline. Should these be `float `instead of `double` ?

nlisker replied:
I was wondering about it myself, but all the other values are `double`s that are cast to `float`s. Wouldn't it also be odd to have the API properties `DoubleProperty` and the peer to use `float`s? Isn't it just a matter of where the cast happens?

kevinrushforth replied:
Yes, it's just a matter of where the cast happens. While we aren't entirely consistent, I think you'll find more places where the peers hold floats rather than double (one notable exception is transforms where we need the extra precision). In particular, Color and position values for shapes are stored in the peer as floats. Also, PhongMaterial stores the power as a float even though the API is double.

nlisker replied:
Alright, I don't mind it, I'll cast in the peer level.

arapte commented on modules/javafx.graphics/src/main/native-prism-d3d/D3DMeshView.cc:
```diff
@@ -154,19 +159,41 @@ void D3DMeshView::render() {
+        lightsAttenuation[a++] = lights[i].attenuation[0];
+        lightsAttenuation[a++] = lights[i].attenuation[1];
+        lightsAttenuation[a++] = lights[i].attenuation[2];
+        lightsAttenuation[a++] = 0;
```
In the initial commit this filed was used for `maxRange`. I think that was good idea. Was there any issue with that approach ?

nlisker replied:
`maxRange` has its own array to reserve space for `minRange` if we ever do that (see [this discussion](https://github.com/javafxports/openjdk-jfx/issues/256#issuecomment-469970201)), so I future-proofed it.

nlisker commented on modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGPointLight.java:
```diff
@@ -26,11 +26,85 @@
+    private static final double DEFAULT_QA = 0;
+    /** Max range default value */
+    private static final double DEFAULT_MAX_RANGE = Double.POSITIVE_INFINITY;
+
```
If they are in `PointLight` it will be difficult for the peer to access them (will need to do through the accessor). Also, the parent `NGLightBase` holds the defaults for `color` and `lightOn`.

nlisker commented on modules/javafx.graphics/src/main/java/javafx/scene/PointLight.java:
```diff
@@ -26,18 +26,35 @@
+ * A light source that radiates light equally in all directions away from itself. The location of the light
+ * source is a single point in space. Any pixel within the range of the light will be illuminated by it,
+ * unless it belongs to a {@code Shape3D} outside of its {@code scope}.
+ * <p>
```
The issue is that range and attenuation work on the pixel scale, not on the node/shape scale. A node can be partially illuminated if only part of it is within the range of the light. See the image in the comment above.

kevinrushforth replied:
Right. This needs to talk about pixels. Perhaps there is a way to make it more clear that we are talking about pixels that are part of a rendered Shape3D, but I don't have a good suggestion right now.

nlisker replied:
Maybe

```
A light source that radiates light equally in all directions away from itself. The location of the light
source is a single point in space. The light affects {@code Shape3D}s in its {@code scope}. Any
pixels in the light's {@code range} that belong to a {@code Shape3D} will be illuminated by it according to the
computation specified in {@link PhongMaterial}.
```
The docs of `PhongMaterial` will need need to be updated too.

kevinrushforth replied:
Yes, I think that change to the docs looks good.

kevinrushforth commented:
> I still need to test your sample app on Mac.

I get the error with your sample app. It fails on Mac or Linux (Ubuntu 16.04) with the same error as I reported above.

nlisker commented:
The error was for the cases of 2 and 3 lights (I was testing 1) and should be fixed now. My fault with copy-paste... that's why we use loops, but I guess this is some optimization for the pipeline. I wonder if it's really woth it over a single shader looping over the number of lights like d3d does.

nlisker commented:
> This will need a fair bit of testing to ensure that there are no regressions either in functionality or (especially) performance, in addition to tests for the new functionality.

Which tests for the new functionality should I write? Up to the shader itself it's mostly just passing on variables, and the API is standard `DoubleProperty`s. I can test the dirty bits / redraw logic.

> On the performance aspect, the inner loop of the lighting calculation now has an additional if test for the max range and additional arithmetic calculations for the attenuation. What we will need is a test program that we can run on Mac and Windows to measure the performance of rendering in a fill-rate-limited case. Ideally, we would not pay much of a performance hit in the default case where `ca == 1, la == 0, qa == 0`, but we first need to be able to measure the drop in performance before we can say whether it is acceptable.

Can you point me to a similar performance test? I didn't see a way to easily measure FPS.
I don't know how to avoid the `if` test for the `maxRange`, I'm open to suggestions. The only thing I can think of is if `maxRange` is infinite (the default), we will swap the shader for one that doesn't make that check. However, swapping shaders also costs performance.

kevinrushforth commented:
We have a few performance tests in apps/performance, but I don't know how up to date they are. They might give you a starting point on how to measure FPS, but really the harder part is going to be coming up with a workload -- a scene with a number of Shape3Ds with large triangles (so that they are fill-rate limited) and 1-3 lights, etc -- that you can use to measure rendering performance without measuring overhead. Typically you want a scene that is rendering continuously in the < 30 fps range, and more like 10 fps to minimize the overhead even more.

Before we figure out whether we need to double the number of shaders (which was one of the ideas that I had as well), we need to know how much of a problem it is. Is it < 2% performance drop on typical cases on a variety of machines or it is a 25% drop (or more likely somewhere in between). If the perf drop is negligible, then it isn't worth doubling the shaders. If it is significant, then we probably need to.

If we do need to double the shaders, I wouldn't do it based on the maxRange being infinite, rather I would do it based on whether attenuation is being used. That way the existing shaders would be unchanged, while the new shader would deal with both attenuation and the maxRange test.

Hopefully, there won't be enough of a perf hit to require doubling the shaders, but we need to be able to measure it.

For functional testing, in addition to the simple API tests, we want to make sure that the basic rendering is working with and without attenuation. Some sort of visual test where you verify that attenuation is / isn't happening as well as testing the cutoff. I wouldn't get too fancy with these...just a sanity test.

kevinrushforth commented:
/reviewers 2

kevinrushforth commented:
/csr

kevinrushforth commented:
Looks like the `jcheck` bot removed the `rfr` label because the CSR isn't complete. An incomplete CSR should be treated the same way as an insufficient number of reviewers. I filed [SKARA-262](https://bugs.openjdk.java.net/browse/SKARA-262) to track this.

nlisker commented:
I've taken the sample app and enlarged the box to fill the whole range of the lights in an attempt to have many pixels rendered for few vertices. I measured 90-115 fps during continuous animation with this patch, and 100-120 before it. I measured the fps using an external app called BandiCam.

Will do more tests this week.

nlisker commented:
Attaching an attempt at testing performance with and without attenuation. Launch `LightingSample`  without the patch or `AttenLightingSample` with the patch. There are 2 modes, one with a single large box and another with many small boxes. Not only is there no difference in performance with and without attenuation, if the lights are turned off (so there is no shader calculation even), there is still no difference. The "single" mode gives ~120 fps, while the "multiple" mode gives ~42 fps and seems to be limited by the number of nodes in the scene rather than anything related to lighting.

Either the calculations for 3 lights is negligible, or the testing is flawed. I tested on Win10 with an RX 470 4GB. Used BandiCam to measure the fps.

[attenTest.zip](https://github.com/openjdk/jfx/files/4238855/attenTest.zip)

nlisker commented:
@kevinrushforth Can you have a look at the test app? I would like to get this moving so we would have time to get the rest of the lighting enhancements into 15.

kevinrushforth commented:
I'll take a look. My quick thought is that we need some sort of test with a reasonable number of large boxes (so that it is fill-limited). If there isn't such a case, and the 3D rendering is always node-limited, then the shader performance doesn't really matter all that much. I suspect we should be able to find a case where it does, but we'll see.

kevinrushforth commented:
I did some limited testing today with a modification to the test program you attached to create a MeshView with 200 large quads (400 triangles) in a single node. This will eliminate the node overhead. I can confirm that it is fill rate limited, because when I send the exact same amount of data, but make the triangles small, the frame rate goes up as expected.

It sill looks like it isn't shader limited, though, at least on my Windows 10 machine, which has an Intel UHD Graphics 630. More testing is needed on other platforms. I'll share the mods to the test program when I have time, but it's basically just creating a set of quads on top of each other by reusing the same 4 points in each pair of faces.

kevinrushforth commented:
I'll attach the above modified testcase that I ran. I ran it on a relatively new Windows 10 laptop and a rather ancient MacBook Pro. I had to drastically reduce the number of quads on the Mac, but the results are similar: no significant difference between the current code and the proposed patch for point lights (without attenuation).

I'd like to see results on a recent machine with a graphics accelerator (either NVIDIA or AMD/ATI) to see if the new shader hurts performance there, but I suspect it will be fine.

kevinrushforth commented:
Updated test case: [attenTest2.zip](https://github.com/openjdk/jfx/files/4332937/attenTest2.zip)

nlisker commented:
On Win 10 with an AMD RX 470 4GB I get the following:

Without the patch:
200 quads average 113 fps
5000 quads average 11.5 fps

With the patch:
200 quads average 106 fps fps
5000 quads average 8.5 fps

Will test on Ubuntu later.

nlisker commented:
On Ubuntu 18 with an AMD RX 470 4GB I get the following:

Without the patch:
200 quads average 107 fps
5000 quads average 11.5 fps

With the patch:
200 quads average 107 fps fps
5000 quads average 11 fps

nlisker commented:
@arapte Can you please test the performance changes too?

kevinrushforth commented:
I think @arapte has a similar MacBookPro model to mine.

I think @prrace might be able to test it (I'll sync with him offline).

kevinrushforth commented:
Here are the results on Phil's machine, which is a Mac Book Pro with a graphics accelerator (Nvidia, I think).

Without the patch:
2000 quads average 8.805 fps

With the patch:
2000 quads average 4.719 fps

Almost a 2x performance hit.

kevinrushforth commented:
Conclusion: The new shaders that support attenuation don't seem to have much of a performance impact on machines with an Intel HD, but on systems with a graphics accelerator, it is a significant slowdown.

So we are left with the two choices of doubling the number of shaders (that is, a set of shaders with attenuation and a set without) or living with the performance hit (which will only be a problem on machines with a dedicated graphics accelerator for highly fill-limited scenes). The only way we can justify a 2x drop in performance is if we are fairly certain that this is a corner case, and thus unlikely to hit real applications.

If we do end up deciding to replicate the shaders, I don't think it is all that much work. I'm more worried about how well it would scale to subsequent improvements, although we could easily decide that for, say, spotlights attenuation is so common that you wouldn't create a version that doesn't do that.

In the D3D HLSL shaders, ifdefs are used, so the work would be to restore the original code and add the new code under an ifdef. Then double the number of lines of gradle (at that point, I'd do it in a for-each loop), then modify the logic that loads the shaders to pick the right one.

For GLSL, the different parts of the shader are in different files, so it's a matter of creating new versions of each of the three lighting shaders that handle attenuation and choosing the right one at runtime.

nlisker commented:
I discussed this with a graphics engineer. He said that a couple of branches do not have any real performance impact even on modern mobile devices, and that, e.g., on iOS 7 using half floats instead of floats was improving shader execution dramatically. Desktops with NVIDIA or AMD and even Intel modern cards can process dozens of branches with no significant performance degradation.

He suggested actually to have all the light types in a single shader file (looking ahead here). He also suggested not to permute on shaders based on the number of lights and just pass in a uniform for that number and loop over it. The permutations on the bump, specular and self illuminations components are correct (not sure we are not doing that for the diffuse component). If we add later shadows, which is not on my near to-do list, then we should permute there.

It also depends on our target hardware. If we take into account hardware from, say, 2005 then maybe branching will cause significant performance loss, but that hinders our ability to increase performance for newer hardware. What is the policy here?

I have a Win10 laptop with a GeForce 610M that I will test this weekend to see if the mobile NVidia cards have some issue.

kevinrushforth commented:
I think most of those are good suggestions going forward. As for the performance drop, the only place we've seen it so far is on graphics accelerators that are a few years old by now. Integrated graphics chipsets (such as Intel HD) either old or new seem largely unaffected by the shader changes. What we are missing is performance metrics from newer graphics accelerators on Mac and Windows.

Even with the performance drop on older graphics devices, I'm leaning towards not having the shaders to be shaders to be doubled, since this is an artificial stress test with huge quads. If we could get performance data from a couple more recent graphics accelerators that would be best.

kevinrushforth commented:
Here is a slightly modified test program. It fixes a compilation error in the previous, and also adds a system property to set the number of quads:

It creates 200 quads by default. If you need to increase this or decrease it to get something in the ~ 10 fps range you can do that with `-DnumQuads=NNNN`.

[pointlighttest.zip](https://github.com/openjdk/jfx/files/4526179/pointlighttest.zip)

prrace commented:
@kevinrushforth
Member
kevinrushforth commented Apr 18, 2020

I think most of those are good suggestions going forward. As for the performance drop, the only place we've seen it so far is on graphics accelerators that are a few years old by now. 

So 50% drop on a 2015 macbook pro is OK ? Do we have numbers on recent macbook pros ?

kevinrushforth commented:
If this were an even remotely representative use case, then no, the performance hit would not be OK. The test was designed as an artificial "worst-case" stress test: a single mesh with a large number of very large (window-sized) quads stacked on top of each other. Any real-world use case won't do this.

We should make sure that we aren't seeing any significant performance drop when rendering spheres (at a couple different tessellation levels) or boxes.

dsgrieve commented:
Results with NVIDIA Quadro P400: 

Without the fix, 1000 quads, average FPS ~7.4
With the fix, 1000 quads, average FPS ~6.1

nlisker commented:
> We should make sure that we aren't seeing any significant performance drop when rendering spheres (at a couple different tessellation levels) or boxes.

I missed this. Do you mean that the test should create a mesh of a sphere instead of a flat surface?

kevinrushforth commented:
I would say in addition to rather than instead of, since both are useful.

What might help is to add the sphere test plus the pathological test I put together into your test program so we can select between them. And then get a few of us to run that updated program and post results.

nlisker commented:
I've added sphere with 10, 64 (default) and 200 divisions. Note that this number is approximate because the sphere mesh corrects the division number with `newDivisions = ((oldDivisions + 3) / 4) * 4`. [attenTest.zip](https://github.com/openjdk/jfx/files/4855989/attenTest.zip)

Using AMD RX 470 4GB on Win 10, all 3 spheres averaged 120 fps both before and after the patch. (It takes a couple of measurements to hit the consistent fps, so I discarded them.)

kevinrushforth commented:
> If you want these commits to be preserved when you integrate, you will need to make a 'merge-style' pull request.

Please ignore this information from Skara. I don't know what caused it, but I've pinged the Skara folks about it.

kevinrushforth commented:
> I've added sphere with 10, 64 (default) and 200 divisions. 

Thanks, I'll take a look at it.

nlisker commented:
Using AMD RX 470 4GB on Ubuntu 18, all 3 spheres averaged 60 fps both before and after the patch.

I wonder why the number is half what it is on Windows. Is it capped?

kevinrushforth commented:
Yes, in general frame rate should be limited to the refresh rate of the monitor, although it isn't in all cases. I would not count on running faster than 60 fps.

kevinrushforth commented on modules/javafx.graphics/src/main/java/com/sun/javafx/scene/PointLightHelper.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2016, 2019, Oracle and/or its affiliates. All rights reserved.
```
Update the "last modified year" for 2020 or else revert (this applies to all copyright year updates).

nlisker replied:
Can the script not get it?

kevinrushforth replied:
Yes, it can. So it would be fine to leave it.

kevinrushforth commented on modules/javafx.graphics/src/main/java/com/sun/javafx/scene/PointLightHelper.java:
```diff
@@ -66,7 +72,6 @@ public static void setPointLightAccessor(final PointLightAccessor newAccessor) {
-
-}
-
+}
```
Minor: missing newline here

kevinrushforth commented on modules/javafx.graphics/src/main/java/com/sun/javafx/sg/prism/NGShape3D.java:
```diff
@@ -173,7 +181,10 @@ private void renderMeshView(Graphics g) {
+                meshView.setPointLight(pointLightIdx++,
+                        0, 0, 0, // x y z
+                        0, 0, 0, 0, // r g b w
+                        1, 0, 0, 0); // ca la qa maxRange
```
Minor: maybe use the getDefaultXxxx methods of NGPointLight?

nlisker replied:
The only impact this has is that the range will be maximal instead of 0. When these reach the shader, they will run the lighting computation as opposed to skipping it. I'm not sure if this will have any performance impact though.

kevinrushforth replied:
In that case, it seems like a generally useful optimization (not just at initialization) to send down `maxRange` as 0 whenever `ca`, `la`, and `qa` are all at their default values.

nlisker replied:
These lights are fillers for the native lights array in the case where the user set less than 3 lights, they are not supposed to do anything, and their color is black. A range of 0 will guarantee that the computation in the shader will be skipped.

In a WIP that I have locally I have removed the "3 lights constraint" altogether, so this part will change in that proposed patch anyway.

kevinrushforth commented on modules/javafx.graphics/src/main/java/javafx/scene/PointLight.java:
```diff
@@ -26,18 +26,35 @@
+ * at a given point in space is {@code color * attn}. It is possible, albeit unrealistic, to specify negative
+ * values to attenuation coefficients. This allows the resulting attenuation factor to be negative, which
+ * results in the light's color being subtracted from the material instead of added to it, thus creating a
+ * "shadow caster".
```
Can you think of any problems that might arise by supporting negative coefficients? If not, then this seems fine.

nlisker replied:
No, I've tested the functionality and it seemed to be working well.

kevinrushforth commented on modules/javafx.graphics/src/main/java/javafx/scene/PointLight.java:
```diff
@@ -70,10 +92,141 @@ public PointLight(Color color) {
+     * node is known to always be outside of the light's range, it is more performant to exclude it from its scope.
+     *
+     * @defaultValue {@code Double.POSITIVE_INFINITY}
+     * @since 14
```
`@since 16`

nlisker commented on modules/javafx.graphics/src/main/resources/com/sun/prism/es2/glsl/main3Lights.frag:
```diff
@@ -68,33 +70,47 @@ varying vec4 lightTangentSpacePositions[3];
+    float dist = length(lightTangentSpacePositions[0].xyz);
+    if (dist <= maxRange) {
+        vec3 l = normalize(lightTangentSpacePositions[0].xyz);
+        float att = 1.0 / (lights[0].attn.x + lights[0].attn.y * dist + lights[0].attn.z * dist * dist);
+        d += clamp(dot(n,l), 0.0, 1.0) * (lights[0].color).rgb * att;
+        s += pow(clamp(dot(-refl, l), 0.0, 1.0), power) * lights[0].color.rgb * att;
```
Interesting, I'd think the compiler would have optimized it.

arapte replied:
The optimizations done by `glCompileShader()` is implementation specific and so little gray area (There may be a way to compare the compiled shaders but I have not investigated on that). But the similar change in hlsl shader is saving one multiplication instruction. So I am assuming that this change might also save just one multiplication instruction. As with this change there is no difference in FPS, it seems trivial. But given it is in fragment shader it is worth saving instructions. I will leave it to you to make a call on this(for both hlsl and glsl). Just that we should keep similar arithmetics in both glsl and hlsl shaders. So either change both or none.

nlisker commented:
@arapte 
> I verified it in the compiled D3D shader, that the change reduces one multiplication instruction in most of the shaders.

I'd like to know how you did that.

arapte commented:
> I'd like to know how you did that.

Compiled hlsl shader binaries are stored in folder `rt\modules\javafx.graphics\build\headers\PrismD3D\hlsl`. The `fxc` command that generates these files is in `build.gradle`.
We can just make a copy and do a text comparison of the generated shader binaries with and without a change.
Note: Existing files should to be deleted to generate new ones.(gradle clean is not necessary)

nlisker commented:
> 1. The API docs look good. If you change the public API to `@since 16` then you can also update the CSR and move it to the "Submitted" state.

I moved it to the PROPOSED state.

> 2. I think it would be good to cleanup the performance test and make it part of this PR, maybe in `tests/performance` (which currently only has the not-very-useful `VMPerformance` subdir) or `tests/manual`.

I added it under `tests/performance`.

> 3. We need some functional tests, ideally automated ones in `tests/system`

What should this test do? Which example should I follow?

By the way, The `tests` project is broken in Eclipse because several source folders, like `tests\system\src\testapp6`, define their own `module-info.java`, and it's illegal in Eclipse to have several of these in a project. To have the tests there compile, the Eclipse files will need to be updated.

kevinrushforth commented:
Given that we don't have any automated tests for lighting (we have a couple that verify that we can take a snapshot and get the same result, but that isn't testing the lighting itself), probably the most we can expect is a simple test of a large quad with a light fairly close to the object, such that the difference with / without attenuation is noticeable. Then you could sample the center and near the corners of the object for both the attenuated and unattenuated cases.

kevinrushforth commented:
The performance tests need a standard GPLv2+classpath copyright header. I haven't tested them yet, but will do that next week.

kevinrushforth commented on modules/javafx.graphics/src/main/native-prism-d3d/hlsl/psMath.h:
```diff
@@ -63,9 +63,17 @@ void phong(
+
+            float3 l = normalize(L[i].xyz);
+            d += saturate(dot(n, l)) * gLightColor[i].xyz * attn;
+            s += pow(saturate(dot(-refl, l)), power) * gLightColor[i].xyz * attn;
```
> ```
> float attenuatedColor = gLightColor[i].xyz / (ca + la * dist + qa * dist * dist);
> ```

Typo: that should be:

```
float3 attenuatedColor = gLightColor[i].xyz / (ca + la * dist + qa * dist * dist);
```

kevinrushforth commented on modules/javafx.graphics/src/main/resources/com/sun/prism/es2/glsl/main1Light.frag:
```diff
@@ -88,9 +88,9 @@ void main()
-        float att = 1.0 / (lights[0].attn.x + lights[0].attn.y * dist + lights[0].attn.z * dist * dist);
-        d = clamp(dot(n, l), 0.0, 1.0) * (lights[0].color).rgb * att;
-        s = pow(clamp(dot(-refl, l), 0.0, 1.0), power) * lights[0].color.rgb * att;
+        float attenuatedColor = (lights[0].color).rgb / (lights[0].attn.x + lights[0].attn.y * dist + lights[0].attn.z * dist * dist);
```
This fails (on macOS) at shader compilation time with:

```
Shader compile log: ERROR: 0:325: Incompatible types in initialization (and no available implicit conversion)
ERROR: 0:326: Use of undeclared identifier 'attenuatedColor'
...
```

That should be `vec3` (same for the other two shader files)

nlisker commented:
I've written a simple draft for the automated test:

        PointLight light = new PointLight(Color.BLUE);
        light.setTranslateZ(-50);
        var box = new Box(100, 100, 1);
        var root = new Group(light, box);
        var scene = new Scene(root);
        var image = box.snapshot(null, null);
        var nonAttenColor = image.getPixelReader().getColor(1, 1);
        light.setLinearAttenuation(2);
        image = box.snapshot(null, null);
        var attenColor = image.getPixelReader().getColor(1, 1);
        assertTrue("Attenuation color should be less than non-attenuated", nonAttenColor.getBlue() > attenColor.getBlue());

However, I'm getting the error:

```
java.lang.UnsupportedOperationException
	at javafx.graphics/test.com.sun.javafx.pgstub.StubToolkit.renderToImage(StubToolkit.java:719)
	at javafx.graphics/javafx.scene.Scene.doSnapshotTile(Scene.java:1383)
	at javafx.graphics/javafx.scene.Scene.doSnapshot(Scene.java:1319)
	at javafx.graphics/javafx.scene.Node.doSnapshot(Node.java:2136)
	at javafx.graphics/javafx.scene.Node.snapshot(Node.java:2214)
```
Where is the test that takes the snapshot? I am missing something in my setup.

kevinrushforth commented:
I guess you are trying a unit test in the `javafx.graphics` project? That won't work. It needs to be a system test in `tests/system/...`

nlisker commented:
I tried several approaches. When I put it under `tests/system` and run the tests task, I don't see it being run. I'm not familiar with the way tests under `tests` are run in general.

kevinrushforth commented:
You need to pass a flag to enable system tests (and a second one for Robot tests, but if you use snapshot it might not need to be).

```
gradle -PFULL_TEST=true -PUSE_ROBOT=true :systemTests:test --tests MyTest
```

nlisker commented:
> gradle -PFULL_TEST=true -PUSE_ROBOT=true :systemTests:test --tests MyTest

What format is `MyTest`? Is it some relative path?

kevinrushforth commented:
It can either be a fully qualified package name (with `.` as separator) or the unqualified name of the test class. The class name must end with exactly `Test`. So for example, try it with `Snapshot1Test`.

nlisker commented:
I wrote the following test:

```
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import javafx.application.Application;
import javafx.application.Platform;
import javafx.scene.Group;
import javafx.scene.PointLight;
import javafx.scene.Scene;
import javafx.scene.paint.Color;
import javafx.scene.shape.Box;
import javafx.stage.Stage;
import javafx.stage.WindowEvent;

public class PointLightAttenuationTest {

    private static CountDownLatch startupLatch;
    private static Stage stage;
    private static PointLight light = new PointLight(Color.BLUE);
    private static Box box = new Box(100, 100, 1);

    @BeforeClass
    public static void initFX() throws Exception {
        startupLatch = new CountDownLatch(1);
        new Thread(() -> Application.launch(TestApp.class, (String[])null)).start();
        assertTrue("Timeout waiting for FX runtime to start", startupLatch.await(15, TimeUnit.SECONDS));
    }

    public class TestApp extends Application {

        @Override
        public void start(Stage mainStage) {
            stage = mainStage;
            light.setTranslateZ(-50);
            var root = new Group(light, box);
            var scene = new Scene(root);
            stage.setScene(scene);
            stage.setFullScreen(true);
            stage.addEventHandler(WindowEvent.WINDOW_SHOWN, e -> Platform.runLater(startupLatch::countDown));
            stage.show();
        }
    }
    
    @Test
    public void testAttenuation() {
        var image = box.snapshot(null, null);
        var nonAttenColor = image.getPixelReader().getColor(1, 1);

        light.setLinearAttenuation(2);
        image = box.snapshot(null, null);
        var attenColor = image.getPixelReader().getColor(1, 1);

        System.out.println(nonAttenColor);
        System.out.println(attenColor);
        if (nonAttenColor.getBlue() > attenColor.getBlue()) {
            throw new AssertionError("Attenuation color should be less than non-attenuated");
        }
    }

    @AfterClass
    public static void teardown() {
        Platform.runLater(() -> {
            stage.hide();
            Platform.exit();
        });
    }
}
```
But when executing it with

    ./gradlew -PFULL_TEST=true -PUSE_ROBOT=true :systemTests:test --tests PointLightAttenuationTest

I get the error

```
test.javafx.scene.lighting3D.PointLightAttenuationTest > classMethod FAILED
    java.lang.AssertionError: Timeout waiting for FX runtime to start
        at org.junit.Assert.fail(Assert.java:91)
        at org.junit.Assert.assertTrue(Assert.java:43)
        at test.javafx.scene.lighting3D.PointLightAttenuationTest.initFX(PointLightAttenuationTest.java:59)
```

So for some reason the Application doesn't start, it seems. I ran `ShapeViewOrderLeakTest` and `RestoreSceneSizeTest` which look the same, and those work. Any idea?

kevinrushforth commented:
If you run it with `--info` you will see something like this:

```
test.javafx.scene.shape.PointLightAttenuationTest STANDARD_ERROR
    Exception in Application constructor
    Exception in thread "Thread-4" java.lang.RuntimeException: Unable to construct Application instance: class test.javafx.scene.shape.PointLightAttenuationTest$TestApp
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:890)
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
        at java.base/java.lang.Thread.run(Thread.java:832)
    Caused by: java.lang.NoSuchMethodException: test.javafx.scene.shape.PointLightAttenuationTest$TestApp.<init>()
        at java.base/java.lang.Class.getConstructor0(Class.java:3427)
        at java.base/java.lang.Class.getConstructor(Class.java:2165)
        at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$8(LauncherImpl.java:801)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:455)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
        at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
        at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
        at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
        ... 1 more
```

The nested `TestApp` class should be declared as `static`. Btw, I don't recommend using `setFullScreen` for a test such as this.

nlisker commented:
Thanks, the test runs now, but I've run into an issue with `snapshot`.

`Node#snapshot` uses its scene parameters (like lights) to render the image, but if the node is in a subscene then the image will be wrong since it takes the wrong data. Bounds transforms like `sceneToLocal` and `localToScene` do take care of this case. I think that this is a mistake in `snapshot`, and maybe it should use the subscene, or a `boolean subScene` parameter should be offered, or `SnapshotParameters` should allow to specify it.

nlisker commented:
Looks like there's more to it. I've created the test code:

```
import javafx.application.Application;
import javafx.scene.Group;
import javafx.scene.PerspectiveCamera;
import javafx.scene.PointLight;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.Slider;
import javafx.scene.image.ImageView;
import javafx.scene.layout.BorderPane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Box;
import javafx.stage.Stage;

public class PointLightAttenuationTest extends Application {

    public static void main(String[] args) throws Exception {
        launch(args);
    }

    @Override
    public void start(Stage stage) {
        var light = new PointLight(Color.BLUE);
        light.setTranslateZ(-10);
        Box box = new Box(100, 100, 1);
        var group = new Group(light, box);

        var scene = new Scene(group, 500, 500, true);
        var camera = new PerspectiveCamera(true);
        camera.setTranslateZ(-300);
        camera.setFarClip(1000);
        scene.setCamera(camera);
        stage.setScene(scene);
        stage.show();

        var imageView = new ImageView();
        var snapshotButton = new Button("Node Snaphot");
        snapshotButton.setOnAction(e -> imageView.setImage(scene.snapshot(null)));

        var slider = new Slider(0, 0.2, 0);
        slider.setShowTickMarks(true);
        slider.setShowTickLabels(true);
        light.linearAttenuationProperty().bindBidirectional(slider.valueProperty());

        var snapshotStage = new Stage();
        snapshotStage.setScene(new Scene(new BorderPane(imageView, snapshotButton, null, slider, null)));
        snapshotStage.setWidth(600);
        snapshotStage.setHeight(600);
        snapshotStage.show();
    }
}
```
If the line `snapshotButton.setOnAction(e -> imageView.setImage(scene.snapshot(null)));`  is replaced to use `box.snapshot(null, null)` then the snapshot is wrong.

kevinrushforth commented:
@nlisker Since there is a possible bug in snapshot, you might consider using Robot screen capture instead?

Also, I don't know if you noticed, but there are now whitespace errors in the shader files after the fix for [JDK-8240499](https://bugs.openjdk.java.net/browse/JDK-8240499).

kevinrushforth reviewed (commented):
The public API and implementation looks good. I left a few inline comments.

I'll review the CSR in parallel with your fixing the few things I noted.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/lighting3D/PointLightAttenuationTest.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    @BeforeClass
+    public static void initFX() throws Exception {
```
One more thing I forgot to mention. You will need to ship the test if 3D is not supported using `assumeTrue`. See [TriangleMeshValidationTest.java#L64](https://github.com/openjdk/jfx/blob/master/tests/system/src/test/java/test/robot/test3d/TriangleMeshValidationTest.java#L64).

nlisker commented on modules/javafx.graphics/src/main/native-prism-d3d/D3DLight.cc:
```diff
@@ -43,6 +43,10 @@ D3DLight::D3DLight() {
+    attenuation[0] = 1;
+    attenuation[1] = 0;
+    attenuation[2] = 0;
+    maxRange = 0;
```
Yes, these values are always overridden.

nlisker commented on modules/javafx.graphics/src/main/native-prism-d3d/D3DLight.cc:
```diff
@@ -56,3 +60,7 @@ void D3DLight::setPosition(float x, float y, float z) {
     position[2] = z;
 }
+
+/*void D3DLight::setRange(float r) {
```
I forgot to mention this point. These setter functions for color and position (and range) are never used since whenever there is a change in the java side the whole array and lights are recreated instead of being adjusted for the change. I'm not familiar enough with the memory model of JNI, but it seems expensive to re-render everything on every change (I think that the mesh is also recreated every time in the native code). Is this the way it's supposed to work?

kevinrushforth replied:
If the lights array is being recreated on a change, there might be a small savings to be had to reuse the existing array (it would need to be updated only during the sync while the renderer is idle).

If the mesh is always being recreated, there may be an opportunity for an even bigger savings, presuming that only the mesh data changed (and not the index values in the faces array).

This could be a future optimization, but we would need something to quantify the possible savings.

kevinrushforth commented on tests/system/src/test/java/test/javafx/scene/lighting3D/PointLightAttenuationTest.java:
```diff
@@ -0,0 +1,140 @@
+
+    @AfterClass
+    public static void teardown() {
+        Platform.runLater(() -> {
```
I forgot to mention that the `@AfterClass` method is still run even if the call to `assumeTrue` in the `@BeforeClass` method causes the test to be skipped. You will need to repeat the check (or set a flag in `initFX` and check that flag) here and skip the cleanup if the initialization was skipped (using an `if` check, not repeating the `assumeTrue`). Otherwise the exception that is thrown here will cause the test to fail.

You can test this by forcing the SW pipeline. I did it like this:

```
_JAVA_OPTIONS="-Dprism.order=sw" gradle --info -PFULL_TEST=true :systemTests:test --tests PointLightAttenuationTest
```

nlisker replied:
The test fails even with this addition. I found that I need to use `@Before` to run the check.

kevinrushforth commented on tests/performance/3DLighting/attenuation/LightingSample.java:
```diff
@@ -0,0 +1,133 @@
+        var sphere = new Button("Sphere");
+        sphere.setOnAction(e -> switchTo(environment.createSphere((int) subdivisionSlider.getValue())));
+
+        var quadSlider = new Slider(500, 10_000, 1000);
```
I was doing some final testing, and ran this on a few different systems. I think the min and max values are too large (especially the min value). I recommend something more like min=100 and max=5000.

kevinrushforth commented on tests/performance/3DLighting/attenuation/LightingSample.java:
```diff
@@ -0,0 +1,133 @@
+    public void start(Stage stage) throws Exception {
+        environment.setStyle("-fx-background-color: teal");
+
+        var subdivisionSlider = new Slider(10, 200, 60);
```
I might recommend setting the max value of the slider to something like 1000, since even on a slow system, it runs quite nicely at that tessellation.

nlisker replied:
I forgot to reply to this and the other comment. Since we're going to implement more light types, this small test app is going to be updated anyway, so I will change the values then.

kevinrushforth reviewed (approved):
I think this is ready to go in once it gets a second reviewer and once the CSR is approved.

I left two very minor comments on the test. I don't mind whether you fix them now or as a follow-up.

arapte reviewed (approved):
Looks good to me. There seems to be an existing issue with PointLight. It is reported here [JDK-8255015](https://bugs.openjdk.java.net/browse/JDK-8255015)

nlisker commented:
/integrate
